keras_nlp/models/llama/llama_attention.py

```diff
@@ -0,0 +1,220 @@
+        )
+
+        query_rot = self.rotary_embedding_layer(query_rot)
+        key_rot = self.rotary_embedding_layer(key_rot)
+
+        query = ops.concatenate((query_rot, query_pass), axis=-1)
+        key = ops.concatenate((key_rot, key_pass), axis=-1)
+
+        print("query shapes = ", ops.shape(query), query.shape)
+        print("key shapes = ", ops.shape(key), key.shape)
+
+        key = self._repeat_kv(key, self.num_key_value_groups)
+        value = self._repeat_kv(value, self.num_key_value_groups)
+
+        attention_output = self._compute_attention(query, key, value, attention_mask)
+
+        attention_output = ops.reshape(
+            attention_output,
+            [
+                ops.shape(attention_output)[0],
+                ops.shape(attention_output)[1],
+                self.hidden_dim,
+            ],
+        )
+
+        attention_output = self._output_dense(attention_output)
+
+        return attention_output
+
+    def _get_common_kwargs_for_sublayer(self, use_bias=True):
+        common_kwargs = {}
+
+        kernel_initializer = clone_initializer(self._kernel_initializer)
+        bias_initializer = clone_initializer(self._bias_initializer)
+
+        common_kwargs["kernel_initializer"] = kernel_initializer
+        if use_bias:
+            common_kwargs["bias_initializer"] = bias_initializer
+
+        return common_kwargs
+
+    def _masked_softmax(self, attention_scores, attention_mask=None):
+        if attention_mask is not None:
+            mask_expansion_axis = -3
+            for _ in range(
+                    len(attention_scores.shape) - len(attention_mask.shape)
+            ):
+                attention_mask = ops.expand_dims(
+                    attention_mask, axis=mask_expansion_axis
+                )
+        return self._softmax(attention_scores, attention_mask)
+
+    def _compute_attention(
+            self, query, key, value, attention_mask=None, training=None
+    ):
+        attention_scores = ops.einsum("aecd,abcd->acbe", key, query)
+
+        norm_factor = ops.sqrt(
+            ops.convert_to_tensor(self.attn_head_size, self.compute_dtype)
+        )
+
+        attention_scores /= norm_factor
+
+        attention_scores = self._masked_softmax(
+            attention_scores, attention_mask
+        )
+        attention_scores = self._attn_dropout_layer(
+            attention_scores, training=training
+        )
+        attention_output = ops.einsum(
+            "acbe,aecd->abcd", attention_scores, value
+        )
+
+        return attention_output
+
+    def _repeat_kv(self, hidden_states, num_repeats):
+        if num_repeats == 1:
+            return hidden_states
+
+        batch_size, seq_len, num_key_value_heads, head_dim = hidden_states.shape
+
+        hidden_states = ops.expand_dims(hidden_states, axis=2)
+        hidden_states = ops.repeat(hidden_states, num_repeats, axis=2)
+
+        repeated_hidden_states = ops.reshape(hidden_states,
```

Comment:
Backbone fails to instantiate. 
```
hidden_dim = 64
num_heads = 4
num_key_value_heads = 2
num_layers = 4
intermediate_dim = hidden_dim // num_heads

backbone = LlamaBackbone(
    vocabulary_size=10,
    num_layers=num_layers,
    num_heads=num_heads,
    num_key_value_heads = num_key_value_heads,
    hidden_dim=hidden_dim,
    intermediate_dim=intermediate_dim,
    max_sequence_length=10,
)

input_batch = {
    "token_ids": ops.ones((2, 5)),
    "padding_mask": ops.ones((2, 5)),
}
```

<details><summary>Details</summary>
<p>
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
[<ipython-input-104-8c02074549d2>](https://localhost:8080/#) in <cell line: 14>()
     12 
     13 
---> 14 backbone = LlamaBackbone(
     15     vocabulary_size=10,
     16     num_layers=num_layers,

6 frames
[/usr/local/lib/python3.10/dist-packages/keras_core/src/backend/tensorflow/numpy.py](https://localhost:8080/#) in tf__reshape(x, new_shape)
     10                 try:
     11                     do_return = True
---> 12                     retval_ = ag__.converted_call(ag__.ld(tfnp).reshape, (ag__.ld(x), ag__.ld(new_shape)), None, fscope)
     13                 except:
     14                     do_return = False

TypeError: Exception encountered when calling layer "transformer_layer_0" (type LlamaDecoder).

in user code:

    File "<ipython-input-35-1ef95ef15f3c>", line 121, in call  *
        x = self._self_attention_layer(
    File "/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py", line 70, in error_handler  **
        raise e.with_traceback(filtered_tb) from None
    File "/tmp/__autograph_generated_file2d2segzs.py", line 21, in tf__call
        key = ag__.converted_call(ag__.ld(self)._repeat_kv, (ag__.ld(key), ag__.ld(self).num_key_value_groups), None, fscope)
    File "/tmp/__autograph_generated_file0izabk6t.py", line 47, in tf___repeat_kv
        ag__.if_stmt(ag__.ld(num_repeats) == 1, if_body, else_body, get_state, set_state, ('do_return', 'retval_', 'hidden_states'), 2)
    File "/tmp/__autograph_generated_file0izabk6t.py", line 34, in else_body
        repeated_hidden_states = ag__.converted_call(ag__.ld(ops).reshape, (ag__.ld(hidden_states), (ag__.ld(batch_size), ag__.ld(seq_len), ag__.ld(num_key_value_heads) * ag__.ld(num_repeats), ag__.ld(head_dim))), None, fscope)
    File "/tmp/__autograph_generated_file3onkdgs5.py", line 12, in tf__reshape
        retval_ = ag__.converted_call(ag__.ld(tfnp).reshape, (ag__.ld(x), ag__.ld(new_shape)), None, fscope)

    TypeError: Exception encountered when calling layer 'llama_attention' (type LlamaAttention).
    
    in user code:
    
        File "<ipython-input-99-eec46ec56a27>", line 125, in call  *
            key = self._repeat_kv(key, self.num_key_value_groups)
        File "<ipython-input-99-eec46ec56a27>", line 202, in _repeat_kv  *
            repeated_hidden_states = ops.reshape(hidden_states, (batch_size, seq_len, num_key_value_heads * num_repeats, head_dim))
        File "/usr/local/lib/python3.10/dist-packages/keras_core/src/backend/tensorflow/numpy.py", line 482, in reshape  *
            return tfnp.reshape(x, new_shape)
    
        TypeError: Failed to convert elements of (None, None, 4, 16) to Tensor. Consider casting elements to a supported type. See https://www.tensorflow.org/api_docs/python/tf/dtypes for supported TF dtypes.
    
    
    Call arguments received by layer 'llama_attention' (type LlamaAttention):
      • hidden_states=tf.Tensor(shape=(None, None, 64), dtype=float32)
      • attention_mask=tf.Tensor(shape=(None, None, None), dtype=int32)


Call arguments received by layer "transformer_layer_0" (type LlamaDecoder):
  • decoder_sequence=tf.Tensor(shape=(None, None, 64), dtype=float32)
  • decoder_padding_mask=tf.Tensor(shape=(None, None), dtype=int32)
  • decoder_attention_mask=None

</p>
</details> 

`hidden_states` shape = (None, None, 2, 2, 16)

How should I apply reshape here?